rtrlib/transport/tcp/tcp_transport.c

```diff
@@ -91,13 +136,65 @@ int tr_tcp_open(void *tr_socket)
 				TCP_DBG("Socket bind failed, %s", tcp_socket, strerror(errno));
 				goto end;
 			}
+
+			freeaddrinfo(bind_addrinfo);
+			bind_addrinfo = NULL;
 		}
 
-		if (connect(tcp_socket->socket, res->ai_addr, res->ai_addrlen) == -1) {
+		if (set_socket_non_blocking(tcp_socket->socket) == TR_ERROR) {
+			TCP_DBG("Could not set socket to non blocking, %s", tcp_socket, strerror(errno));
+			goto end;
+		}
+
+		if (connect(tcp_socket->socket, res->ai_addr, res->ai_addrlen) == -1 && errno != EINPROGRESS) {
 			TCP_DBG("Couldn't establish TCP connection, %s",
 				tcp_socket, strerror(errno));
 			goto end;
 		}
+
+		freeaddrinfo(res);
+		res = NULL;
+
+		fd_set wfds;
+
+		FD_ZERO(&wfds);
+		FD_SET(tcp_socket->socket, &wfds);
+		int oldcancelstate;
+
+		/* Enable cancalability for the select call
+		 * to prevent rtr_stop from blocking for a long time.
+		 * It must be the only blocking call in this function.
+		 * Since local resources have all been freed this should be safe.
+		 */
+		pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, &oldcancelstate);
+		int ret = select(tcp_socket->socket + 1, NULL, &wfds, NULL, NULL);
+
+		pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &oldcancelstate);
+
+		if (ret < 0) {
+			TCP_DBG("Could not select tcp socket, %s", tcp_socket, strerror(errno));
+			goto end;
+
+		} else if (ret == 0) {
```

Comment:
can this actually happen, as there is no timeout given in `select` above?

Reply:
I'm wondering why I used select without a timeout there. I do not remember a specific reason and it does seem wrong to me right now.
I think I will just add a timeout like I have in the ssh implementation.